idxmgmt/ilm/config.go

```diff
@@ -18,34 +18,117 @@
+	Overwrite     bool        `config:"overwrite"`
+	RequirePolicy bool        `config:"require_policy"`
+	Policies      []Policy
+}
+
+//Policy binds together an ILM policy's name and body with an event type
+type Policy struct {
+	Policy    map[string]interface{}
+	EventType string
+	Name      string
 }
 
-// Enabled indicates whether or not ILM should be enabled
+//Enabled indicates whether or not ILM should be enabled
 func (c *Config) Enabled() bool {
 	return c.Mode != libilm.ModeDisabled
 }
 
-// ModeString stringifies enabled mode
-// This is a workaround as strings should be generated in libbeat.
-func ModeString(m libilm.Mode) string {
-	switch m {
-	case libilm.ModeAuto:
-		return "auto"
-	case libilm.ModeEnabled:
-		return "true"
-	default:
-		return "false"
+//NewConfig returns an ILM config, where default configuration is merged with user configuration
+func NewConfig(ucfg *common.Config) (Config, error) {
+	policies := policyPool()
+	policyMappings := policyMapping()
+	c := defaultConfig()
+	if ucfg != nil {
+		if err := ucfg.Unpack(&c); err != nil {
+			return Config{}, err
+		}
+
+		var cfg config
+		if err := ucfg.Unpack(&cfg); err != nil {
+			return Config{}, err
+		}
+		// create a collection of default and configured policies
+		for name, policy := range cfg.Policies {
+			policies[name] = prepare(policy)
+		}
+		//update policy name per event according to configuration
+		for _, entry := range cfg.Mapping {
+			if _, ok := policyMappings[entry.Event]; !ok {
+				return c, errors.Errorf("event_type '%s' not supported for ILM setup", entry.Event)
+			}
+			policyMappings[entry.Event] = entry.Policy
+		}
+	}
+
+	for event, policyName := range policyMappings {
+		policy, ok := policies[policyName]
+		if !ok {
+			if c.RequirePolicy {
+				return Config{}, errors.Errorf("policy '%s' not configured for ILM setup", policyName)
+			}
+			policy = nil
+		}
+		c.Policies = append(c.Policies, Policy{EventType: event, Policy: policy, Name: policyName})
+	}
+	return c, nil
+}
+
+func defaultConfig() Config {
+	return Config{Mode: libilm.ModeAuto, Overwrite: false, RequirePolicy: true}
+}
+
+type config struct {
+	Mapping []struct {
+		Policy string `config:"policy"`
+		Event  string `config:"event_type"`
+	} `config:"mapping"`
+	Policies map[string]policy `config:"policies"`
+}
+
+type policies map[string]policy
+type policy map[string]interface{}
+
+var errPolicyFmt = errors.New("input for ILM policies is in wrong format")
+
+func (p *policies) Unpack(i interface{}) error {
+	inp, ok := i.(map[string]interface{})
+	if !ok {
+		return errPolicyFmt
+	}
+	*p = map[string]policy{}
+	for k, v := range inp {
+		inpP, ok := v.(map[string]interface{})
+		if !ok {
+			return errPolicyFmt
+		}
+		(*p)[k] = policy(inpP)
+	}
+	return nil
+}
+
+//prepare ensures maps are in the format elasticsearch expects for policy bodies,
+//it replaces nil values with an empty map
+func prepare(bb map[string]interface{}) map[string]interface{} {
```

Comment:
I struggle to understand the need for this. Can't you do all the error handling you want in the `Unpack`?
As a matter of fact, how do I craft an `apm-server.yml` so that the result is different with and without the `prepare` function?

Reply:
When defining something like `delete: {}` in the yml file, the way it is resolved in code is `delete: nil`. This is not an accepted format by Elasticsearch, and does not reflect what is configured, so I need to ensure it is properly mapped to have an empty hash instead of nil. 
I followed your suggestion and moved it into the `Unpack` method.